### DISCUSSION TITLE
'undefined' not considered JSON serializable?

### DISCUSSION THREAD

**Discussion by dBianchii** [liveblocks/liveblocks#1501] [Help] [answered]
- I have read this article: https://liveblocks.io/docs/platform/troubleshooting#lson-constraint-and-interfaces

Hi, I have a specifc type for my chess game that I need to convert from a normal useState() to a liveblocks storage state. The type I need to store is this:

```ts
type Piece = {
  id: keyof typeof pieceIds;
};
type Board = (Piece | undefined)[][];
//     ^  This I need!!!
```

When I try to use it, I am doing something like this:
![image](https://github.com/liveblocks/liveblocks/assets/17372417/b8fa596f-3465-4a58-9810-13f2824853b8)

As you can see, when I use it with `null` there are no errors. However, when I try to set it to undefined, it says it doesn't meet the constraint for a `Lson`.

I thought that undefined was JSON serializable. What's going on? Do I really need to convert my type to `null`? I don't want that....

**Comment by nvie** [comment, marked as answer]
Unfortunately, JavaScript does not allow `undefined` values to be serialized in JSON arrays. An array like `[undefined]` would become `[null]` on the other end, due to the fact that `JSON.parse(JSON.stringify([undefined])` returns `[null]`. This is unfortunately out of our control. We therefore decided to not allow `undefined` values as list members, because this implicit conversion could take people by surprise.

Hope that helps to explain it.
